classes/Configuration/XdmodConfiguration.php

```
@@ -0,0 +1,224 @@
+            $children = array_keys($extends);
+
+            // finding the extended that do not themselves extend something. aka the bottom of the
+            // dependency tree.
+            $roots = array_diff($targets, $children);
+
+            // finding the everything that extends these root elements.
+            $extenders = array_keys(
+                array_filter(
+                    $extends,
+                    function ($key) use ($roots) {
+                        return in_array($key, $roots);
+                    }
+                )
+            );
+
+            // Actually do the extending.
+            $this->resolveExtends($source, $extenders);
+
+            // Now that we've finished extending $extenders, remove them so we don't process them
+            // again
+            foreach ($extenders as $extender) {
+                unset($extends[$extender]);
+            }
+        }
+    }
+
+    /**
+     * Iterate through $config and find any objects that contain an `extends` property for later
+     * processing.
+     *
+     * The returned array will be in the form:
+     * array(
+     *   'extender' => 'extendee',
+     *    ...
+     * )
+     *
+     * a concrete example would be:
+     * array(
+     *   'usr' => 'default', // usr extends default
+     *   'pi'  => 'usr',     // pi extends usr
+     *   'cs'  => 'pi',      // cs extends pi
+     *   'cd'  => 'cs'       // cd extends cs
+     * )
+     *
+     * @param stdClass $config the object that is being iterated over.
+     * @param string|null $parentKey the key of the object that contains `extends`.
+     * @return array in the form:
+     * array(
+     *   'extender' => 'extendee'
+     * )
+     */
+    protected function findExtends(\stdClass $config, $parentKey = null)
+    {
+        $extends = array();
+        foreach ($config as $k => $v) {
+            if ($k === self::EXTENDS_KEYWORD && is_string($v)) {
+                $extends[$parentKey] = $v;
```

What happens when we set `$extends[null] = $v`?

---

So this one's an interesting one, I've done some local testing which I'll detail below:

For the scenario that I believe you're referencing, I ran the following:
```php
<?php 

function findExtends(\stdClass $config, $parentKey = null)
{
    $extends = array();
    foreach ($config as $k => $v) {
        if ($k === 'extends' && is_string($v)) {
            $extends[$parentKey] = $v;
        } elseif (is_object($v)) {
            $extends = array_merge($extends, findExtends($v, $k));
        }
    }
    return $extends;
}

$empty = new stdClass();
$empty->extends = 'default';

print_r( findExtends($empty) );
```

and I ended up with: 
```
Array
(
    [] => default
)
```

After plugging in `$this->resolveExtends($source, $extenders);` and the code to support it ( I wanted to see if this code would exception out when it was passed the malformed data ). I ended up with: 
```
stdClass Object
(
    [extends] => default
)
```

So I think my slightly longer answer to the implied question of, how does the code react when $parentKey is null and it encounters an 'extends' keyword? would be that, it goes along its merry old way and processes things just fine. I don't think that the user / admin who has the funktastic  configuration files is going to get the behavior they expect. But the code won't fall over, at least at this point, if they do.

---

I could come up with a quick json schema that describes, generically, how our config files should be setup, so that we can throw a quick error pre-processing if they have something funky going on.

---

I just want to be sure that we don't end up with some sort of strange or undefined behavior if `findExtends()` gets called somewhere without the 2nd argument such as in `handleExtendsFor()`